File: largeda/largeda.cpp
#include "nix.hpp"
#include <iostream>
#include <time.h>

using namespace nix;

std::vector<float> run(size_t N) {
    std::vector<float> times(N);
    nix::File file = nix::File::open("/tmp/data-benchmark.nix", nix::FileMode::Overwrite);
    nix::Block b = file.createBlock("test", "test");
    std::vector<double> data = std::vector<double>(N);
    for (size_t i = 0; i < N; i++) {
        std::string name = "times" + nix::util::numToStr(i);
        nix::DataArray da = b.createDataArray(name, "nix.event.positions", nix::DataType::Double, {1});
        const clock_t begin_time = clock();
        da.setData(data);
        times[i] = float(clock () - begin_time) /  CLOCKS_PER_SEC;
    }
    file.close();
    return times;
}


int main(int argc, char** argv){
    size_t N = atoi(argv[1]);
    std::vector<float> times = run(N);
    for (size_t i = 0; i < N; i++) {
        std::cout << i+1 << ", " << times[i] << std::endl;
    }
    return 0;
}

File: largeda/largeda.py
import h5py
from time import time
import numpy as np


def runtest(nixfile, N):
    blk = nixfile.create_block("blk", "blk")
    times = []
    for n in range(N):
        data = np.random.random(n)
        name = "data" + str(n)
        # blk.create_data_array(name, "test", data=data)
        da = blk.create_data_array(name, "test", shape=(n,))
        t0 = time()  # measure write data time only!
        da.write_direct(data)
        t1 = time()
        times.append(t1-t0)
    return times


def benchmarkdirect(N):
    h5file = h5py.File("/tmp/append-benchmark-direct.nix", "w")
    blk = h5file.create_group("blk")
    times = []
    for n in range(N):
        data = np.random.random(n)
        name = "data" + str(n)
        t0 = time()
        da = blk.create_group(name)
        dataset = da.require_dataset(
            name, shape=(n,), dtype=data.dtype,
            chunks=True, maxshape=(None,)
        )
        dataset[:] = data
        t1 = time()
        times.append(t1-t0)
    h5file.close()
    return times
